std/math/cmp/generic.go

```diff
@@ -0,0 +1,91 @@
+package cmp
+
+import (
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/std/math/bits"
+	"math/big"
+)
+
+// IsLessUnsigned returns 1 if a < b, and returns 0 if a >= b. The comparison
+// will be unsigned and all field elements will be treated as positive numbers.
+// Therefore, If the underlying field is of prime order P, the elements will be
+// considered to represent integers in range [0, P-1].
+func IsLessUnsigned(api frontend.API, a, b frontend.Variable) frontend.Variable {
```

Comment:
Variables are field elements and canonical integer representation is in `[0, P-1]`. I think repeating that the methods are unsigned is redundant and potentially misleading. I would rename just to `IsLess` and `IsLessOrEqual` below.

Reply:
Hey @aybehrouz I am having issues due to exactly the unsigned comparison issue. Here is me discussing the issue : https://github.com/ConsenSys/gnark/discussions/705 . Can you help me understand how should I deal with these negative integers?

Reply:
The proof system uses a finite field of prime order P. (i.e. it has P elements) Since prime fields are isomorphic to modular arithmetic, we can assume that our arithmetic is performed modulus P. The elements of modular arithmetic are equivalence classes and all numbers of the form `x + k*P` belong to the same equivalence class. (all have the same remainder when divided by P) Modular arithmetic is like ordinary integer arithmetic but numbers map into equivalence classes. Therefore, you can use it for any arithmetic as long as you can **uniquely** determine your numbers from these equivalence classes.

For signed arithmetic, the most natural way is to have the same number of positive and negative numbers, so we assume that our numbers are in range [-P/2,P/2]. Then, we will have to make sure that the result of any operation is never less than -P/2 or greater that P/2. (I'm not trying to state boundary cases accurately here) 

gnark internally uses positive numbers for representing equivalence classes, and all number are shown in range [0, P-1]. Therefore, -1 is shown as P-1, -2 as P-2, etc... `api.Cmp` also assumes the same representation, so if you want to check that a number is negative, you should check if it is in range [P/2, P-1]. (boundaries are not accurate)

Reply:
Thanks a lot @aybehrouz . How do I fix the value of P based on my inputs or is it fixed internally? if the later, then how do I know the value of P to be able to do this?

Reply:
You can query the value of P by `api.Compiler().Field()`.

Reply:
Also be careful about overflows:
 > Then, we will have to make sure that the result of any operation is never less than -P/2 or greater that P/2.

Reply:
Thanks a lot. This helped immensely.

Reply:
How does this conversion thing work with ```api.Div```? In my circuit I am dividing an odd number by 2. So the result is a decimal, resulting in big number and constraint failing because I don't know what to write in ```AssertIsEqual ```.

Reply:
Division works by using the multiplicative inverse. If the result of your division is not an integer you'll get weird results. For example if P=11 the multiplicative inverse of 2 would be 6. Therefore, division by 2 works by multiplying into 6. If you want to divide 4 by 2, it would be `4 * 6 mod 11` which is 2. However, if you divide 3 by 2 you'll get `3 * 6 mod 11` which is 7.

I strongly recommend that you consider using a ZK-VM for your task, instead of using gnark. gnark is a low level library for creating compact ZK circuits. You'll need to be familiar with ZK proof systems, and you can easily create insecure circuits. A ZK-VM on the other hand provides a convenient tool for verifying arbitrary computation without the need for digging into the details.

Reply:
Thanks that makes a lot of sense. I did not know about ZK-VMs. Can I please have some links for some good/common ZKVMs? Why don't people use ZK-VMs often? or do they? I am assuming I would have to write the computations as functions in Golang and then feed the functions to ZK-VMs? Is that how they work?
Also this is for an academic project where I need exact time needed to verify and lengths of proofs. So I would need compact circuits, I guess the ZK-VMs will have redundancy in the circuits.

Reply:
> Also this is for an academic project where I need exact time needed to verify and lengths of proofs. So I would need compact circuits, I guess the ZK-VMs will have redundancy in the circuits.

Oh, if that's the case, then you can't use a ZK-VM. You'd have to use a low level library, like gnark, for creating specialized circuits. A VM always uses the same circuit for every computation.

Reply:
Thanks!